blib/man3/Simo.3
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Simo 3"
.TH Simo 3 "2008-12-16" "perl v5.10.0" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
Simo \- Very simple framework for Object Oriented Perl.
.SH "VERSION"
.IX Header "VERSION"
Version 0.01
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Sh "Creating class and accessors"
.IX Subsection "Creating class and accessors"
Simo is framework that simplify Object Oriented Perl.
If you use Simo, You can write class and accessors very very simple way.
.PP
If you want to class and accessors, you do this way.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&    
\&    # accessors
\&    sub title{ ac }
\&    sub author{ ac }
\&    sub description{ ac }
.Ve
.PP
This code create Book class and three accessors, title,author and description.
You no longer write accessors implementations repeat.
.PP
and
.PP
new method is automatically created.
You no longer write new method.
.Sh "Using class and accessors"
.IX Subsection "Using class and accessors"
You use this class this way.
.PP
.Vb 1
\&    use Book;
\&    
\&    # constructor
\&    my $book = Book\->new(
\&        title => \*(AqHappy book\*(Aq,
\&        author => \*(AqKen\*(Aq,
\&        description => \*(Aqthis give you happy\*(Aq,
\&    );
\&    
\&    # get value
\&    $book\->title;
\&    
\&    # set value
\&    $book\->author( \*(AqTaro\*(Aq );
.Ve
.PP
You pass args to new method. args is key-value pair.
If you pass key which is not defined as accessor, the script will die.
.PP
If you get value, you write \f(CW$book\fR\->title;
.PP
If you set value, you write \f(CW$book\fR\->title( 'Taro' );
.Sh "Define default value"
.IX Subsection "Define default value"
You can define default value for accessor.You do this way.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&
\&    sub title{ ac( \*(AqPapa\*(Aq ) }
.Ve
.PP
It is also very simple to define default value. 
You can pass default value as ac function first arg.
.Sh "Value check by regular expression"
.IX Subsection "Value check by regular expression"
You can check value passed to accessor using regular expression.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&
\&    sub price{ ac( 2000, qr/^\ed+$/ ) }
\&    
\&    package main;
\&    use Book;
\&    my $book = Book\->new;
\&    $book\->price( \*(Aqaaaa\*(Aq ) # this code die.
.Ve
.PP
You can pass regular expression reference( qr// ) as ac function second arg.
.PP
Note: default value is not checked. in this sample, 2000 is not checked.
.Sh "Hook function for validation or filter."
.IX Subsection "Hook function for validation or filter."
You can define hook function for accessor.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&
\&    sub date{ ac( \*(Aq20080101\*(Aq, \e&date_filter ) }
\&    
\&    #filter ( for example, \*(Aq2008\-10\-10\*(Aq \-> \*(Aq20081010\*(Aq )
\&    sub date_filter{
\&        my ( $val ) = @_;
\&        $val =~ s/\-//g;
\&        return $val
\&    }
\&    
\&    package main;
\&    use Book;
\&    my $book = Book\->new;
\&    $book\->date( \*(Aq2008\-10\-10\*(Aq );
\&    print $book\->date; # 20081010
.Ve
.PP
You can pass code reference as ac function second arg.
The code reference is executed before set value by accessor.
.PP
In this example, If you set '2008/10/10', It change '20081010'.
.PP
Your hook subroutine can receive two args.
.PP
1. passed value by setter.
.PP
2. self ( this object )
.PP
If you need, you can access self.
.PP
.Vb 2
\&    sub date_filter{
\&        my ( $val, $self ) = @_;
\&        
\&        $val =~ s/\-//g;
\&        $self\->title; # You can call other accessors or methods.
\&        ...
\&    }
.Ve
.SH "Automatically type convert"
.IX Header "Automatically type convert"
If you pass array to accessor, its array convert to array ref.
    \f(CW$book\fR\->title( 'a', 'b' );
    \f(CW$book\fR\->title; # get [ 'a', 'b' ], not ( 'a', 'b' )
.SH "Class option"
.IX Header "Class option"
Simo has class options. You set options this way.
.PP
.Vb 1
\&    use Simo qw( option1 option2 ... );
.Ve
.Sh "arg_top option"
.IX Subsection "arg_top option"
If you set arg_top option, first arg of new passed specified accessor. 
if args count is odd number
.PP
.Vb 1
\&    use Simo qw( arg_top:title );
\&
\&    my $book = Book\->new( \*(AqHappy book\*(Aq, author => \*(AqKen\*(Aq );
\&                          # first arg passed title accessor
\&                          # if args count is odd number
.Ve
.Sh "Array wrapper"
.IX Subsection "Array wrapper"
You can wrap array using your array wrapper class
.PP
.Vb 3
\&    # use your list wrapper class
\&    package Book;
\&    use Simo qw( ary_struct:Book::Array );
\&
\&    sub authors{ ac };
\&    
\&    
\&    package Book::Array; 
\&    sub new{
\&        bless $_[1], ref( $_[0] ) || $_[0]; 
\&    }
\&    
\&    sub length{
\&        my $self = shift;
\&        return scalar @{ $self }
\&    }
\&    
\&    or
\&    
\&    # use list wrapper class in CPAN
\&    package Book;
\&    use List::Rubyish; 
\&    
\&    use Simo qw( ary_struct:List::Rubyish );
\&    sub authors{ ac };
.Ve
.PP
You can use this class
.PP
.Vb 3
\&    use Book;
\&    my $book = Book\->new;
\&    $book\->authors( \*(AqKen\*(Aq, \*(AqTom\*(Aq ); #set authors
\&    
\&    print $book\->authors\->length; # get and call length method. result is 2.
.Ve
.Sh "Hash wrapper"
.IX Subsection "Hash wrapper"
You can also use hash_struct otpion
.PP
.Vb 2
\&    package Book;
\&    use Simo qw( hash_struct:Book::Hash );
.Ve
.Sh "Setter return value type"
.IX Subsection "Setter return value type"
You can set setret option this way.
.PP
.Vb 2
\&    package Book;
\&    use Simo qw( setret:before );
.Ve
.PP
If you use setret option You can select setter return value type
.PP
1.current : return value is current value
.PP
.Vb 2
\&    $book\->title( 1 );
\&    my $ret = $book\->title( 2 ); # $ret is 2
.Ve
.PP
2.before : return value is before value
.PP
.Vb 2
\&    $book\->title( 1 );
\&    my $ret = $book\->title( 2 ); # $ret is 1
.Ve
.PP
3.chain : return value is self
.PP
.Vb 2
\&    $book\->title( 1 );
\&    my $ret = $book\->title( 2 ); # $ret is $book
\&
\&    # so you can use method chain.
\&    $book\->title( \*(Aqa\*(Aq )\->author( \*(Aqb\*(Aq );
.Ve
.SH "Accessor option"
.IX Header "Accessor option"
You can use accessor option.
.PP
You can set many option more than default value and hook function.
.PP
If you must pass 'opt' as first arg and  pass hash ref as second arg,
.PP
do this way.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&
\&    sub title{ ac( opt => { default => 2, hook => sub{} } ) };
.Ve
.PP
The following is example, all accessor option set.
.PP
.Vb 10
\&    sub title{
\&        ac( opt => {
\&            default => 2,
\&            hook => sub{ },
\&            setret => \*(Aqbefore\*(Aq,
\&            ary_struct => \*(AqBook::Array\*(Aq,
\&            hash_struct => \*(AqBook::Hash\*(Aq,
\&            hash_force => 1,
\&        });
\&    }
.Ve
.PP
Accessors option is same as class options except hook or hash_force;
.PP
If you use accessor option, you can adapt opiton only individual accessor.
.PP
I still do not said to you about hash_force option.
.Sh "force setter to receive list as hash"
.IX Subsection "force setter to receive list as hash"
You can force setter to receive list as hash
.PP
.Vb 3
\&    package Book;
\&    use Simo;
\&    sub country_id_list{ ac( opt => { hash_force => 1 }) }
\&    
\&    package main;
\&    my $book = Book\->new;
\&    $book\->country_id_list( japan => \*(Aqjp\*(Aq, USA => \*(Aqus\*(Aq );
\&         # { japan => \*(Aqjp\*(Aq, USA => \*(Aqus\*(Aq } ,not [ \*(Aqjapan\*(Aq, \*(Aqjp\*(Aq, \*(AqUSA\*(Aq, \*(Aqus\*(Aq]
.Ve
.SH "EXPORT"
.IX Header "EXPORT"
This class exports ac function. you can use ac function to implement accessor.
.SH "FUNCTIONS"
.IX Header "FUNCTIONS"
.Sh "ac"
.IX Subsection "ac"
ac is accessor interface. if you create accessor, you do this.
    package Book;
    use Simo;
.PP
.Vb 2
\&    sub title{ ac }
\&    ...
.Ve
.PP
You can use this method as getter and setter.
.PP
Getter is
    \f(CW$book\fR\->title;
.PP
Setter is 
    \f(CW$book\fR\->title( 'Bird Adventure' );
.Sh "new"
.IX Subsection "new"
new is created automatically. 
by default, new receive hash args.
.PP
.Vb 1
\&    my $book = Book\->new( title => \*(AqPaPa is good\*(Aq, author => \*(AqMaMa\*(Aq );
.Ve
.SH "MORE TECHNIQUES"
.IX Header "MORE TECHNIQUES"
I teach you useful techniques.
.Sh "want to create new function by yourself."
.IX Subsection "want to create new function by yourself."
Please look at '$self\->SUPER::new' in this code.
.PP
What I shoud said to you is that Book extends Simo, 
although there are no 'use base' or '@ISA'.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&
\&    sub new{
\&        my ( $self, @args ) = @_;
\&        # ... You write what you want to do
\&        
\&        return $self\->SUPER::new( @args );
\&    }
\&
\&    sub title{ ac }
.Ve
.PP
you can call SUPER::new method.
.PP
If you want to create new function by yourself,
first you write what you do 
and finally call \f(CW$self\fR\->SUPER::new.
.Sh "inheritance"
.IX Subsection "inheritance"
You can inherit super clsss.
.PP
This sample : PBook extends Book.
.PP
.Vb 2
\&    package Book;
\&    use Simo;
\&
\&    sub title{ ac }
\&
\&    package PBook;
\&    use base \*(AqBook\*(Aq;
\&    use Simo;
\&    
\&    sub author{ ac } # PBook has two accessor, title and author
.Ve
.PP
You can do inheritance by using normal way that use 'use base'.
If you define accessors simply in PBook package,
please use Simo again.
.SH "AUTHOR"
.IX Header "AUTHOR"
Yuki Kimoto, \f(CW\*(C`<kimoto.yuki at gmail.com>\*(C'\fR
.SH "BUGS"
.IX Header "BUGS"
Please report any bugs or feature requests to \f(CW\*(C`bug\-simo at rt.cpan.org\*(C'\fR, or through
the web interface at <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Simo>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.
.SH "SUPPORT"
.IX Header "SUPPORT"
You can find documentation for this module with the perldoc command.
.PP
.Vb 1
\&    perldoc Simo
.Ve
.PP
You can also look for information at:
.IP "\(bu" 4
\&\s-1RT:\s0 \s-1CPAN\s0's request tracker
.Sp
<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Simo>
.IP "\(bu" 4
AnnoCPAN: Annotated \s-1CPAN\s0 documentation
.Sp
<http://annocpan.org/dist/Simo>
.IP "\(bu" 4
\&\s-1CPAN\s0 Ratings
.Sp
<http://cpanratings.perl.org/d/Simo>
.IP "\(bu" 4
Search \s-1CPAN\s0
.Sp
<http://search.cpan.org/dist/Simo/>
.SH "ACKNOWLEDGEMENTS"
.IX Header "ACKNOWLEDGEMENTS"
.SH "COPYRIGHT & LICENSE"
.IX Header "COPYRIGHT & LICENSE"
Copyright 2008 Yuki Kimoto, all rights reserved.
.PP
This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Version 0.0201
    I don't pack test file in package

Version 0.02
    new method can receive hash ref.

Version 0.01_05
    Remove hook accessor option.
    Add set_hook accessor option and get_hook accessor option.
    
    Fix sub _SIMO_ac_opt
    $Simo::info{ class }{ $class }{ ac }{ opt } = $opt;
    to
    $Simo::info{ class }{ $class }{ ac }{ $key }{ opt } = $opt;
    
Version 0.01_04
    fix to be right inherit order 
    
    unshift @{ "${caller_class}::ISA" }, __PACKAGE__;
    to
    push @{ "${caller_class}::ISA" }, __PACKAGE__;

Version 0.01_03
    refactoring
    
    qq/sub ${ac_define_class}::${key}/ . qw/{ _SIMO_ac_real( '$key' , \@_ ) }/;
    to
    qq/sub ${ac_define_class}::${key} { _SIMO_ac_real( '$key' , \@_ ) }/;

Version 0.01_02
    fix this failed test.
    FAIL Simo-0.01 i386-netbsd-thread-multi-64int 4.0

Version 0.01_01
    first developer release



lib/Simo.pm
package Simo;
use strict;
use warnings;
use Carp;

our $VERSION = '0.0201';

sub import{
    my $caller_class = caller;
    
    {
        # export function
        no strict 'refs';
        *{ "${caller_class}::ac" } = \&Simo::ac;
        
        # caller inherit Simo
        push @{ "${caller_class}::ISA" }, __PACKAGE__;
    }

    # auto strict and warnings
    strict->import;
    warnings->import;
}

sub new{
    my ( $class, @args ) = @_;
    
    # check args
    @args = %{ $args[0] } if ref $args[0] eq 'HASH';
    confess 'please pass key value pairs to new method' if @args % 2;
    
    # bless
    my $self = {};
    bless $self, ref $class || $class;
    
    # set args
    while( my ( $key, $val ) = splice( @args, 0, 2 ) ){
        no strict 'refs';
        eval{ $self->$key( $val ) };
        croak $@ if $@;
    }
    return $self;
}

# Accessor register
sub ac(@){

    # accessor info
    my ( $self, $key, $ac_define_class, @vals ) = _SIMO_ac_info();
    
    # check accessor info
    my $class = ref $self;
    confess "Cannot call accessor from Class." unless $class;
    
    # check and rearrange accessor option;
    my $ac_opt = {};
    my %valid_opt = map{ $_ => 1 } qw( default set_hook get_hook hash_force );
        
    $ac_opt->{ default } = shift if @_ % 2;
    
    while( my( $key, $val ) = splice( @_, 0, 2 ) ){
        confess "$key is not valid accessor option" unless $valid_opt{ $key };
        $ac_opt->{ $key } = $val;
    }
    
    # register accessor option
    _SIMO_ac_opt( $ac_define_class, $key, $ac_opt );

    # redefine real acessor
    my $ac_redefine = qq/sub ${ac_define_class}::${key} { _SIMO_ac_real( '$key' , \@_ ) }/;
    
    {
        no warnings 'redefine';
        eval $ac_redefine;
    }
    
    # call accessor
    $self->$key( @vals );
}

# Real accessor.
sub _SIMO_ac_real{
    my ( $key, $self, @vals ) = @_;
    
    # check args
    my $class = ref $self;
    confess "Cannot call accessor from Class." unless $class;
    
    # get accessor defined class
    my $ac_define_class = _SIMO_ac_define_class( $class, $key );
    
    # get accessor option
    my $ac_opt = _SIMO_ac_opt( $ac_define_class, $key );
    
    # rearrange value;
    my $val = @vals == 1 ? $vals[0] :
              @vals >= 2 && $ac_opt->{ hash_force } ? { @vals } :
              @vals >= 2 ? [ @vals ] :
              undef;
    
    # return value( return old_value in case setter is called )
    my $ret = _SIMO_obj_updated($self,$key) ? $self->{$key} : $ac_opt->{default};
    
    # set value if value is defined
    if( defined( $val ) ){
        # setter hook function
        if( $ac_opt->{ set_hook } ){
            eval{ $val = $ac_opt->{ set_hook }->($self,$val) };
            confess $@ if $@;
        }
        
        # set new value
        $self->{ $key } = $val;
        
        # updated is true
        _SIMO_obj_updated( $self, $key, 1 );
    }
    else{
        # getter hook function
        if( $ac_opt->{ get_hook } ){
            eval{ $ret = $ac_opt->{ get_hook }->($self, $ret) };
            confess $@ if $@;
        }
    }
    
    return !wantarray ? $ret :
           ref $ret eq 'ARRAY' ? @{ $ret } :
           ref $ret eq 'HASH' ? %{ $ret } :
           ( $ret );
}

# Get accessor define class
sub _SIMO_ac_define_class{
    my ( $class, $key ) = @_;
    unless( $Simo::info{ class }{ $class }{ ac }{ $key }{ ac_define_class } ){
        
        my $ac_define_class = ( caller 2 )[ 3 ];
        
        if( $ac_define_class =~ /^(.+)::/ ){
            $ac_define_class = $1;
        }
        
        $Simo::info{ class }{ $class }{ ac }{ $key }{ ac_define_class } = $ac_define_class;
    }
    return $Simo::info{ class }{ $class }{ ac }{ $key }{ ac_define_class };
}

# Get and set accessor opt
sub _SIMO_ac_opt{
    my( $class, $key, $opt ) = @_;
    if( defined( $opt ) ){
        $Simo::info{ class }{ $class }{ ac }{ $key }{ opt } = $opt;
    }
    return $Simo::info{ class }{ $class }{ ac }{ $key }{ opt };
}

# Check whether default is updated.
sub _SIMO_obj_updated{
    my ( $self, $key, $val ) = @_;
    if( defined( $val ) ){
        $Simo::info{ obj }{ $self }{ ac }{ $key }{ updated } = $val;
    }
    return $Simo::info{ obj }{ $self }{ ac }{ $key }{ updated };
}

# Helper to get acsessor info;
sub _SIMO_ac_info {
    package DB;
    my @caller = caller 2;
    
    my ( $self, @vals ) = @DB::args;
    my $sub = $caller[ 3 ];
    my ( $ac_define_class, $key ) = $sub =~ /^(.*)::(.+)$/;

    return ( $self, $key, $ac_define_class, @vals );
}

# Simo object destroctor
sub DESTROY{
    my $self = shift;
    delete $Simo::info{ obj }{ $self };
}

=head1 NAME

Simo - Very simple framework for Object Oriented Perl.

=head1 VERSION

Version 0.01

=cut

=head1 FEATURES

Simo is framework that simplify Object Oriented Perl.

The feature is that

=over 4

=item 1.You can write accessors in very simple way.

=item 2.Inheritable new method is prepared.

=item 3.You can set default value and hook function for accessors

=back

If You use Simo, you are free from bitter work 
writing new and accessors repeatedly

=cut

=head1 SYNOPSIS

    package Book;
    use Simo;
    
    # simple accessors
    sub title{ ac }
    
    # having default value
    sub author{ ac 'Kimoto' }
    
    # all accessor option 
    sub description{ ac
        default => 'This is good book',
        set_hook => sub{
            my ( $self, $val ) = @_;
            # do what you want to do
            return $val;
        },
        get_hook => sub{
            my ( $self, $val ) = @_;
            # do what you want to do
            return $val;
        },
        hash_force => 1
    }
    1;

=head1 DESCRIPTION

=head2 Creating class and accessors

You can create class and accessors.

    package Book;
    use Simo;
    
    # accessors
    sub title{ ac }
    sub author{ ac }
    sub description{ ac }
    1;

This sample create Book class and three accessor mothods( title,author and description ),
and new method is automatically created.

=cut

=head2 Using class and accessors

You can use Book class.
    use Book;
    
    # constructor
    my $book = Book->new(
        title => 'Happy book',
        author => 'Ken',
        description => 'this give you happy',
    );
    
    # get value
    my $title = $book->title; # 'Happy book'
    
    # set value
    $book->author( 'Taro' );

You can pass key-value pairs to new method. Key is need to be same name as accessor.

If you pass nonexistent key, script will die.

you can get value :

    $book->title;

you can set value :

    $book->title( 'Taro' );
    
=head2 Get old value

You can get old value when you use accessor as setter.

    $book->author( 'Ken' );
    my $old_value = $book->author( 'Taro' ); # $old_value is 'Ken'

=head2 accessor having Default value

You can set default value for accessor.

    package Book;
    use Simo;

    sub title{ ac 'Papa' }
    
You get 'Papa' if 'title' field is not initialized.

=cut

=head2 Setter Hook function for validation or filter or etc.

You can set set_hook function for accessor.

    package Book;
    use Simo;
    
    # set_hook function for accessor
    sub date{ ac set_hook => \&date_filter }
    
    # set_hook function definition
    sub date_filter{
        my ( $self, $val ) = @_;
        $val =~ s/-//g; # ( '2008-10-10' -> '20081010' )
        return $val;
    }
    1;

If you set date this way
    $book->title( '2008-08-11' );
    
'2008-08-08' is filtered, and change to '20080811'.

=cut

=head2 set_hook function arguments and return value

set_hook function receive two args( $self and $val ).

In this example, $self is $book object.

$val is passed value '2008-08-11'

if you pass array to setter, $val has been converted to array ref.

and you must return scalar, not list. 

=cut

=head2 Getter Hook function for initialization or etc.

You can set get_hook function for accessor.

    package Book;
    use Simo;
    
    # get_hook function for accessor
    sub date{ ac
        get_hook => sub{
            my ( $self, $val ) = @_;
            if( !$val ){ # val is $self->{ 'conf' }
                
                $val = localtime;
                
                $self->date( $val ); # if you set 
            }
            return $val;
        }
    }

    1;

you get date

    $book->date;

if $book->{ 'date' } is undef, $book->date return current localtime.

get_hook option is useful, if you want to set default value in dynamically, not Statically.

=cut

=head2 get_hook function arguments and return value

get_hook function receive two args( $self and $val ).

In this example, $self is $book object.

$val is current value undef.

and you must return scalar, not list. 

=cut

=head2 Automatically type convert

If you pass array to accessor, array convert to array ref.
    $book->title( 'a', 'b' );
    $book->title; # get [ 'a', 'b' ], not ( 'a', 'b' )

=cut

=head2 Accessor option

You can set accessor option.

=head3 default option

You can set default value for accessor.

    sub author{ ac 'Kimoto' }

or explicitely

    sub author{ ac default => 'Kimoto' }

=cut

=head3 set_hook option

You can define hook function for setter

    sub date{ ac set_hook => \&filter }
    
    sub filter{
        my ( $self, $val ) = @_;
        # ...
        return $some_val;
    }
    
or using anonymous function

    sub data{ ac 
        set_hook => sub{
            my( $self, $val ) = @_;
            # ...
            return $some_val;
        }
    }

=cut

=head3 get_hook option

You can define hook function for getter

    sub date{ ac
        get_hook => sub{
            my ( $self, $val ) = @_;
            if( !$val ){ # val is $self->{ 'conf' }
                
                $val = localtime;
                
                $self->date( $val ); # if you set 
            }
            return $val;
        }
    }
=cut

=head3 hash_force option

If you pass array to accessor, Normally list convert to array ref.
    $book->title( 'a' , 'b' ); # convert to [ 'a', 'b' ]

Even if you write
    $book->title( a => 'b' )

( a => 'b' ) converted to [ 'a', 'b' ] 

If you use hash_force option, you convert list to hash ref

    sub country_id{ ac hash_force => 1 }

    $book->title( a => 'b' ); # convert to { a => 'b' }

=cut

=head2 Multiple accessor option setting sample

    # one line
    sub title{ ac 'Pure love', hook => \&filter, hash_force => 1 }
    
    sub filter{
        # ..
    }
    
    # multiple line, hook function is anonimous
    sub title{ ac { k => 1 },
        hash_force => 1,
        hook => sub {
            # ..
        }
    }    
    
    # multiple line, default is explicitely, hook function is anonimous
    sub title{ ac
        default => { k => 1 },
        hash_force => 1,
        hook => sub {
            # ..
        }
    }

=cut

=head2 scalar context and list context

You can call accessor in scalar context and list context.

Accessor is designed to suit context.

You call accessor in scalar context, you get scalar value( scalar, array ref, hash ref, object, etc )

    my $ary = $book->authors; # you get array ref

You calls accessor in list context, you get list.

    my @ary = $book->authors; # you get list

=cut

=head1 EXPORT

This class exports ac function. you can use ac function to implement accessor. 

=head1 FUNCTIONS

=head2 ac

You can define accsessor using ac function.

    package Book;
    use Simo;
    
    sub title{ ac }
    ...

You can use this accessor.

Get is
    $book->title;

Set is 
    $book->title( 'Bird Adventure' );

=cut

=head2 new

New method is created automatically.
it receive key-value pairs as args.

    my $book = Book->new( title => 'PaPa is good', author => 'MaMa' );

=cut

=head1 MORE TECHNIQUES

I teach you useful techniques.

=head2 New method overriding

by default, new method receive key-value pairs.
But you can change this action by overriding new method.

For example, Point class. You want to call new method this way.

    my $point = Point->new( 3, 5 ); # xPos and yPos

You can override new method.
    
    package Point;
    use Simo;

    sub new{
        my ( $self, $x, $y ) = @_; # two arg( not key-value pairs )
        
        # You can do anything if you need
        
        return $self->SUPER::new( x => $x, y => $y );
    }

    sub x{ ac }
    sub y{ ac }
    1;

Simo implement inheritable new method.
Whenever You change argments or add initializetion,
You override new method.

=cut

=head2 Extend base class

you may want to extend base class. It is OK.

But I should say to you that there are one thing you should know.
The order of Inheritance is very important.

I write good sample and bad sample.

    # base class
    package Book;
    sub title{ ac };
    
    # Good sample.
    # inherit base class. It is OK!
    package Magazine;
    use base 'Book'; # use base is first
    use Simo;        # use Simo is second;
    
    # Bad sample
    package Magazine;
    use Simo;          # use Simo is first
    use base 'Book';   # use base is second

If you call new method in Good sample, you call Book::new method.
This is what you wanto to do.

If you call new method in Bad sample, you call Simo::new method. 
you will think why Book::new method is not called?

Maybe, You will be wrong sometime. So I recomend you the following writing.

    package Magazine; use base 'Book'; # package and base class
    use Simo;                          

It is like other language class Definition and I think looking is not bat.
and you are not likely to choose wrong order.

=head1 AUTHOR

Yuki Kimoto, C<< <kimoto.yuki at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-simo at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Simo>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Simo


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Simo>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Simo>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Simo>

=item * Search CPAN

L<http://search.cpan.org/dist/Simo/>

=back


=head1 SEE ALSO

L<Class::Accessor>,L<Class::Accessor::Fast>, L<Moose>, L<Mouse>.

=head1 COPYRIGHT & LICENSE

Copyright 2008 Yuki Kimoto, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Simo
